### ISSUE TITLE
datasets.fetch_atlas_harvard_oxford()  got wrong number of label

### ISSUE THREAD

**Issue by poppintiger** [nilearn/nilearn#1871]
I want to get the ROI timeseries from rest fmri:
`mask_data = datasets.fetch_atlas_harvard_oxford('cort-maxprob-thr25-2mm')
mask = mask_data.maps
labels = mask_data.labels`

There 49 labels in this atlas, but when I do the converting, the label changed from different rest fmri files.

`masker = NiftiLabelsMasker(labels_img=mask, standardize=True,memory='nilearn_cache', verbose=10)
time_series = masker.fit_transform(dataset, confounds=compcor)
print(np.shape(time_series))`

The result turn out to be (150, 46) or (150, 47), but I am supposed to get the right shape （150, 49). Is that also likely because my fmri is not in MNI space?

**Comment by poppintiger** [comment]
[https://pan.baidu.com/s/1Itwy3PwgfEzY0icNlYC8zA](url)
This is my frmi data.

**Comment by kchawla-pi** [comment]
Hi @poppintiger , you can write code blocks by encasing them in triple backticks. That will make this much easier to read.

What is `dataset` and `compcor`? Would you add that to this code snippet?

**Comment by KamalakerDadi** [comment]
@poppintiger for some reason I could not access the data from the links you are providing. 

I suppose you used NiftiLabelsMasker to extract timeseries. I would advice you to change ```resampling_target=labels```. Change this resampling strategy and let us know.

**Comment by poppintiger** [comment]
> compcor
@kchawla-pi  thanks for your reminder!
’‘’
data = datasets.fetch_adhd(n_subjects=1)
dataset = data.func[0]
compcor = high_variance_confounds(dataset, n_confounds=5)

mask_data = datasets.fetch_atlas_harvard_oxford('cort-maxprob-thr25-2mm')
mask = mask_data.maps
labels = mask_data.labels

from nilearn.input_data import NiftiLabelsMasker
masker = NiftiLabelsMasker(labels_img=mask, standardize=True,memory='nilearn_cache', verbose=10)

time_series = masker.fit_transform(dataset, confounds=compcor)
print('timeseries:',np.shape(time_series))
‘’‘
The mask have 49 labels, but the result of my time series only have 48 labels and the number changed from different rest fmri image.

**Comment by poppintiger** [comment]
Hi ，@KamalakerDadi ，I have tried your recommended parameter，but the final time series still lack several labels .

**Comment by KamalakerDadi** [comment]
>The mask have 49 labels, but the result of my time series only have 48 labels and the number changed from different rest fmri image.

Mask meaning atlas. The atlas has 49 labels including background. You should not count background label as a brain region. In that sense, 48 is the right which you have got it.

 You can print the labels the after fetching harvard oxford atlas and indicates that first label is background. NiftiLabelsMasker avoids background label by default.

**Comment by poppintiger** [comment]
> > The mask have 49 labels, but the result of my time series only have 48 labels and the number changed from different rest fmri image.
> 
> Mask meaning atlas. The atlas has 49 labels including background. You should not count background label as a brain region. In that sense, 48 is the right which you have got it.
> 
> You can print the labels the after fetching harvard oxford atlas and indicates that first label is background. NiftiLabelsMasker avoids background label by default.

Thanks @KamalakerDadi:
Indeed the first label is the background, but for some rest fmri image, I got only 46 labels. So where might be wrong with my processing?  Is that possible my image not in MNI space?

**Comment by KamalakerDadi** [comment]
>Indeed the first label is the background, but for some rest fmri image, I got only 46 labels.

Visually check this rest fmri image with others and also check the affine. It might be that not all images have same affine.

**Comment by poppintiger** [comment]
> > Indeed the first label is the background, but for some rest fmri image, I got only 46 labels.
> 
> Visually check this rest fmri image with others and also check the affine. It might be that not all images have same affine.

I got the idea ! Thanks so much   @KamalakerDadi

**Comment by lillepeder** [comment]
Any updates on the issue? 

I have the same issue using FreeSurfer segmentation as a labels_img in ```NiftiLabelsMasker```. I suspect it is because some labels are outside the extent of the EPI file, but I'm not sure. Anyway, this is very troublesome when dealing with multiple subjects - some come out with 167 timeseries, others 168 or 169. In other words I cannot compare them directly. 

I have tried to extract the signal from one region at a time, using ```image.math_img``` in a for loop. This technically works, but this is extremely inefficient (by orders of magnitude).

Is there a way to get the ```labels_img``` index together with its timeseries? This would make my days much easier! 

Any feedback is appreciated.

**Comment by bthirion** [comment]
This is problematic indeed. 
I don't remember having such an issue myself. Can sombedoy share an image/atlas so that we reproduce the issue ?
Thx !

**Comment by KamalakerDadi** [comment]
Me neither. But, I would first use ```resampling_target=labels```. With this you are trying to resample fmri images to atlas. 

Otherwise, I would resample every images i.e., fmri images, mask, atlas to one resolution. This can be done using ```resample_img``` or ```resample_to_img```.

Then, visually check the resampled atlas in particular the number of labels. Then, use ```resampling_target=None``` in ```NiftiLabelsMasker``` if there are no missing labels.

If still problem exists, then I would check with changing ```interpolation``` strategy while resampling offline.

**Comment by jeromedockes** [comment]
are you fitting the masker separately to get different numbers of labels? please check that the number of time series you get corresponds to the length of the labelmasker's `labels_`.

please see 

https://neurostars.org/t/nilearn-niftilabelsmasker-returns-less-timecourses-than-labels-by-more-than-1/3756

**Comment by jeromedockes** [comment]
> troublesome when dealing with multiple subjects - some come out with 167 timeseries, others 168 or 169. In other words I cannot compare them directly.

fit the masker only once and use it on all subjects. or, use the fitted masker's `labels_` attribute to know which timeseries corresponds to which region

**Comment by lillepeder** [comment]
> 
> 
> > troublesome when dealing with multiple subjects - some come out with 167 timeseries, others 168 or 169. In other words I cannot compare them directly.
> 
> fit the masker only once and use it on all subjects. or, use the fitted masker's `labels_` attribute to know which timeseries corresponds to which region

The `labels_ ` attribute seems to be just what I needed. Now it is a trivial task to compare subjects. Much appreciated!

> are you fitting the masker separately to get different numbers of labels?
My data is in native space, using subject-specific segmentations - but if I understood, is it possible to input a list of segmentations and call `fit_transform` all-in-one to a list of EPI files?

Of course it doesn't solve the more general problem that certain regions are left out (i.e. those that are only a few voxels in size), but I think changing my resampling strategy as suggested by @KamalakerDadi would fix that one for future reference :)

**Comment by lillepeder** [comment]
> 
> 
> This is problematic indeed.
> I don't remember having such an issue myself. Can sombedoy share an image/atlas so that we reproduce the issue ?
> Thx !

I will check my permissions to share (patient data is quite strictly regulated).

**Comment by Remi-Gau** [comment]
- I am pretty sure that this issue is related to https://github.com/nilearn/nilearn/issues/3085

So will close it as duplicate.
